### PR TITLE
Remove execution for connection manager

### PR DESCRIPTION
**Pull request by smklein**
- In an attempt to simplify errors, reduce polymorphism, and reduce build times, this PR removes support for "pool-based queries", instead preferring that clients check out connections before sending queries to the underlying pool.

## Previously

- `AsyncConnection` was implemented for both connections and for the pool itself. This meant the query execution functions (i.e., `execute_async(...)`) could act on either a pool or a connection.
- Although this was flexible, it was arguably **too** flexible -- clients would need to deal with `ConnectionError` errors from queries to connections, or `PoolError` errors from queries to pools (which is basically `ConnectionError` + Any error from checking out the connection).

## Now

- `AsyncConnection` is implemented for connections, not pools. This should unify error handling for callers, and hopefully make the access to individual DB connections a little more obvious.

Part of https://github.com/oxidecomputer/omicron/issues/4132

### CODE REVIEW

**Review by davepacheco — APPROVED** [review]
Looks good by itself, though I'd wait for the corresponding Omicron PR to be ready to land before landing this one (in case we find problems there).

Out of curiosity, do other crates impl Connection for the bb8 pool or was that just something we did?

**Comment by smklein** [comment]
> Looks good by itself, though I'd wait for the corresponding Omicron PR to be ready to land before landing this one (in case we find problems there).

Sounds good, I'll wait until presubmit CI goes green.

> Out of curiosity, do other crates impl Connection for the bb8 pool or was that just something we did?

I really **thought** I remembered other crates doing this, but I can't find other good examples where "the pool also implements the connection type".

https://docs.rs/diesel/latest/diesel/prelude/trait.Connection.html is implemented on "Connection" types, that aren't pools. 

I took some inspiration from https://docs.rs/bb8-diesel/ when building this crate, and there, https://docs.rs/bb8-diesel/0.2.1/bb8_diesel/struct.DieselConnectionManager.html does **not** implement the `Connection` trait.

**Pull request merged**
